oauth/client_factory_builder.go

```diff
@@ -0,0 +1,148 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package oauth
+
+import (
+	"fmt"
+	"net/url"
+
+	"github.com/redhat-appstudio/service-provider-integration-operator/api/v1beta1"
+	cli "github.com/redhat-appstudio/service-provider-integration-operator/cmd/oauth/oauthcli"
+	"github.com/redhat-appstudio/service-provider-integration-operator/pkg/spi-shared/httptransport"
+	authz "k8s.io/api/authorization/v1"
+	corev1 "k8s.io/api/core/v1"
+	certutil "k8s.io/client-go/util/cert"
+
+	"net/http"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// K8sClientFactoryBuilder allows to construct different types of K8S client factories
+type K8sClientFactoryBuilder struct {
+	Args cli.OAuthServiceCliArgs
+}
+
+func (r K8sClientFactoryBuilder) CreateInClusterClientFactory() (clientFactory K8sClientFactory, err error) {
+	mapper := meta.NewDefaultRESTMapper([]schema.GroupVersion{})
+	mapper.Add(corev1.SchemeGroupVersion.WithKind("Secret"), meta.RESTScopeNamespace)
+	mapper.Add(v1beta1.GroupVersion.WithKind("SPIAccessTokenDataUpdate"), meta.RESTScopeNamespace)
+	clientOptions, errClientOptions := clientOptions(mapper)
+	if errClientOptions != nil {
+		return nil, errClientOptions
+	}
+	if r.Args.KubeConfig != "" {
+		restConfig, err := clientcmd.BuildConfigFromFlags("", r.Args.KubeConfig)
+		setInsecure(r.Args, restConfig)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create rest configuration: %w", err)
+		}
+		return InClusterK8sClientFactory{ClientOptions: clientOptions, RestConfig: restConfig}, nil
+	}
+	restConfig, err := rest.InClusterConfig()
+	setInsecure(r.Args, restConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize in-cluster config: %w", err)
+	}
+	return InClusterK8sClientFactory{ClientOptions: clientOptions, RestConfig: restConfig}, nil
+}
+
+func (r K8sClientFactoryBuilder) CreateUserAuthClientFactory() (clientFactory K8sClientFactory, err error) {
+	// we can't use the default dynamic rest mapper, because we don't have a token that would enable us to connect
+	// to the cluster just yet. Therefore, we need to list all the resources that we are ever going to query using our
+	// client here thus making the mapper not reach out to the target cluster at all.
+	mapper := meta.NewDefaultRESTMapper([]schema.GroupVersion{})
+	mapper.Add(authz.SchemeGroupVersion.WithKind("SelfSubjectAccessReview"), meta.RESTScopeRoot)
+	mapper.Add(v1beta1.GroupVersion.WithKind("SPIAccessToken"), meta.RESTScopeNamespace)
+	mapper.Add(v1beta1.GroupVersion.WithKind("SPIAccessTokenDataUpdate"), meta.RESTScopeNamespace)
+	clientOptions, errClientOptions := clientOptions(mapper)
+	if errClientOptions != nil {
+		return nil, errClientOptions
+	}
+
+	if r.Args.ApiServer == "" {
```

Comment:
Does it mean that if I have apiserver set, the cluster must have workspaces available?

Reply:
at this moment - yes.. is something may be wrong on that ?

Reply:
it's changing the SPI expectation of cluster setup, it is not documented anywhere that when you set apiserver, you must setup workspaces as well.